enhancements/network/on-prem-configurable-lb-dns-stack.md

```diff
@@ -0,0 +1,332 @@
+
+- First, verify that IP address is valid and kube-api is available using the new IP address.
+- Update CoreDNS to resolve api-int to the new IP address
+- Remove the HAProxy pods after graceful shutdown timeout
+- Keepalived static pods will continue running
+
+-- external to self-hosted Load balancer
+
+To re-activate the self-hosted Load balancer the admin should set apiintIpAddress value to API-VIP address provided in install-config file and verify that external Load balancer is operational for graceful timeout.
+
+Post this change the cluster-hosted-net-services-operator should run the same procedure described above and run the HAProxy pods.
+
+### User Stories
+
+#### As an admin I want to disable the self-hosted on-prem LB and DNS stack
+
+#### As an admin I want to use the self-hosted on-prem LB and DNS stack also in UPI deployments
+
+#### As an admin I want to retrieve the self-hosted stack status through K8S API
+
+### Test Plan
+
+The operator will be tested in the OpenShift CI via e2e tests triggered using the [e2e-metal-ipi](https://github.com/openshift/release/blob/master/ci-operator/step-registry/baremetalds/e2e/baremetalds-e2e-workflow.yaml) workflow defined in the OpenShift CI steps registry and also by e2e tests for openstack, ovirt and vsphere.
+
+Additionally, the following tests should be added:
+
+- Check each platform when external Load balancer is used
+- Disruptive test that switch on/off the self-hosted components
+- Check Kube-api downtime (the values ​​should be similar to the values ​​when using static pods) for master node reboot and node drain cases.
+
+### Upgrade / Downgrade Strategy
+
+Prior to this proposal, the self-hosted stack implementation was running through the MCO.
+The implementation of the self-hostd stack using the cluster-hosted-net-services operator might lead to cases during upgrade/downgrade in which two parallel components are running the self-hosted stack.
+
+A. If the release includes the cluster-hosted-net-services operator that means MCO in this release shouldn't include the HAProxy, CoreDNS and MDNS-Publisher static pods.
+
+B. The cluster-hosted-net-services operator will:
+
+- Have lower priority than MCO, meaning it will be applied before MCO.
+- Use the same code used by MCO to generate configuration files of HAProxy CoreDNS in each node.
+- Run the CoreDNS and HAProxy instances in each node following verification that the corresponding instance created by MCO isn't running anymore in this node (this should be applicable since both static and DaemonSet pod run on host's networking and listen on a well-known port)
+
+A and B will ensure that :
+
+- LB and DNS services will be available even when both MCO and cluster-hosted-net-services operator run the self-hosted stack.
+- Following upgrade to a release includes cluster-hosted-net-services operator only HAProxy ,CoreDNS and MDNS-Publisher DaemonSet pods will run.
+
+## Alternatives
```

Comment:
this needs a convincing discussion why we don't leave the static pods as static pods and just transfer ownership to the new operator. Why do we go with DaemonSets? Which gain let's us accept the risk that DaemonSets are destabilizing the cluster?

Reply:
To be very clear: rewrite this enhancement with static pods, i.e. only the operator changes that creates them from MCO to the new one, and 90% of concerns and questions go away.

Reply:
Enhancement was updated